libs/wire-subsystems/src/Wire/EnterpriseLoginSubsystem/Interpreter.hs

```diff
@@ -0,0 +1,321 @@
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+preAuthorizeImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  new <- case mOld of
+    Nothing -> do
+      let new = DomainRegistration domain PreAuthorized Allowed Nothing
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == None -> do
+      let new = sdr {domainRedirect = PreAuthorized} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == PreAuthorized -> pure sdr
+    _ -> throw EnterpriseLoginSubsystemErrorInvalidDomainRedirect
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/preauthorize"
+  sendAuditMail url "Domain locked" mOld (Just new)
+
+getDomainRegistrationImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r
+  ) =>
+  Domain ->
+  Sem r DomainRegistration
+getDomainRegistrationImpl domain = do
+  mSdr <- tryGetDomainRegistrationImpl domain
+  case mSdr of
+    Just dr -> pure dr
+    Nothing -> throw EnterpriseLoginSubsystemErrorNotFound
+
+tryGetDomainRegistrationImpl ::
+  forall r.
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r
+  ) =>
+  Domain ->
+  Sem r (Maybe DomainRegistration)
+tryGetDomainRegistrationImpl domain = do
+  mSdr <- lookup domain
+  maybe (pure Nothing) (fmap Just . deserialize) mSdr
+  where
+    deserialize :: StoredDomainRegistration -> Sem r DomainRegistration
+    deserialize sdr = do
+      let mDomainRegistration = fromStored sdr
+      case mDomainRegistration of
+        Nothing -> do
+          Log.err $ Log.field "domain" (toByteString' domain) . Log.msg (Log.val "Invalid stored domain registration")
+          throw EnterpriseLoginSubsystemInternalError
+        Just dr -> pure dr
+
+fromStored :: StoredDomainRegistration -> Maybe DomainRegistration
+fromStored sdr =
+  DomainRegistration sdr.domain
+    <$> getDomainRedirect sdr
+    <*> getTeamInvite sdr
+    <*> pure sdr.dnsVerificationToken
+  where
+    getTeamInvite :: StoredDomainRegistration -> Maybe TeamInvite
+    getTeamInvite = \case
+      StoredDomainRegistration _ _ ti _ _ tid _ -> case (ti, tid) of
+        (AllowedTag, _) -> Just Allowed
+        (NotAllowedTag, _) -> Just NotAllowed
+        (TeamTag, Just teamId) -> Just $ Team teamId
+        _ -> Nothing
+
+    getDomainRedirect :: StoredDomainRegistration -> Maybe DomainRedirect
+    getDomainRedirect = \case
+      StoredDomainRegistration _ dr _ ssoId url _ _ -> case (dr, ssoId, url) of
+        (NoneTag, _, _) -> Just None
+        (LockedTag, _, _) -> Just Locked
+        (PreAuthorizedTag, _, _) -> Just PreAuthorized
+        (SSOTag, Just idpId, _) -> Just $ SSO idpId
+        (BackendTag, _, Just beUrl) -> Just $ Backend beUrl
+        (NoRegistrationTag, _, _) -> Just NoRegistration
```

Comment:
here also replace all `_` with `Nothing`.  we should return `DomainRedirect`s from inconsistent `StoredDomainRegistration`s.